satpy/tests/test_demo.py

```diff
@@ -192,3 +192,99 @@ def test_ahi_partial_download(self):
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        return
+
+    def raise_for_status(self):
+        return
+
+    def _get_fake_viirs_sdr_bytesio(self):
+        filelike_obj = io.BytesIO()
+        filelike_obj.write(self._filename.encode("ascii"))
+        filelike_obj.seek(0)
+        return filelike_obj
+
+    def iter_content(self, chunk_size):
+        """Return generator of 'chunk_size' at a time."""
+        bytes_io = self._get_fake_viirs_sdr_bytesio()
+        x = bytes_io.read(chunk_size)
+        while x:
+            yield x
+            x = bytes_io.read(chunk_size)
+
+
+class TestVIIRSSDRDemoDownload:
+    """Test VIIRS SDR downloading."""
+
+    ALL_BANDS = ("I01", "I02", "I03", "I04", "I05",
+                 "M01", "M02", "M03", "M04", "M05", "M06", "M07", "M08", "M09", "M10",
+                 "M11", "M12", "M13", "M14", "M15", "M16",
+                 "SVDNB",
+                 )
+    ALL_GEO = ("GITCO", "GMTCO", "GDNBO")
+
+    @mock.patch('satpy.demo.viirs_sdr.requests')
+    def test_download(self, _requests, tmpdir):
+        """Test downloading and re-downloading VIIRS SDR data."""
+        from satpy.demo import get_viirs_sdr_20170128_1229
+        _requests.get.side_effect = _FakeRequest
+        files = get_viirs_sdr_20170128_1229(base_dir=str(tmpdir))
+        assert len(files) == 10 * (16 + 5 + 1 + 3)  # 10 granules * (5 I bands + 16 M bands + 1 DNB + 3 geolocation)
+        self._assert_bands_in_filenames_and_contents(self.ALL_BANDS + self.ALL_GEO, files, 10)
+
+        get_mock = mock.MagicMock()
+        _requests.get.return_value = get_mock
+        new_files = get_viirs_sdr_20170128_1229(base_dir=str(tmpdir))
+        assert len(new_files) == 10 * (16 + 5 + 1 + 3)  # 10 granules * (5 I bands + 16 M bands + 1 DNB + 3 geolocation)
+        get_mock.assert_not_called()
+        assert new_files == files
+
+    @mock.patch('satpy.demo.viirs_sdr.requests')
+    def test_download_channels_num_granules(self, _requests, tmpdir):
+        """Test downloading and re-downloading VIIRS SDR data."""
+        from satpy.demo import get_viirs_sdr_20170128_1229
+        _requests.get.side_effect = _FakeRequest
+        files = get_viirs_sdr_20170128_1229(base_dir=str(tmpdir),
+                                            channels=("I01", "M01"))
+        assert len(files) == 10 * (1 + 1 + 2)  # 10 granules * (1 I band + 1 M band + 2 geolocation)
+        self._assert_bands_in_filenames_and_contents(("I01", "M01", "GITCO", "GMTCO"), files, 10)
+
+        files = get_viirs_sdr_20170128_1229(base_dir=str(tmpdir),
+                                            channels=("I01", "M01"),
+                                            granules=(2, 3))
+        assert len(files) == 2 * (1 + 1 + 2)  # 2 granules * (1 I band + 1 M band + 2 geolocation)
+        self._assert_bands_in_filenames_and_contents(("I01", "M01", "GITCO", "GMTCO"), files, 2)
```

Comment:
Should there be an assertion making sure the data has not been redownloaded?

Reply:
The other test has a check for this, but yeah maybe I could add that here too.